core/src/main/java/org/apache/calcite/util/Sources.java

```diff
@@ -189,14 +198,43 @@ private static File urlToFile(URL url) {
       return Paths.get(uri).toFile();
     }
 
+    private static URL fileToUrl(File file) {
+      String filePath = file.getPath();
+      if (!file.isAbsolute()) {
+        // convert relative file paths
+        if (File.separatorChar != '/') {
+          filePath = filePath.replace(File.separatorChar, '/');
+        }
+        if (file.isDirectory() && !filePath.endsWith("/")) {
+          filePath += "/";
+        }
+        try {
+          return new URL(
+              "file", null, 0, URLEncoder.encode(filePath, StandardCharsets.UTF_8.name()));
+        } catch (MalformedURLException e) {
+          throw new IllegalArgumentException("Unable to create URI for file " + filePath, e);
```

Comment:
```suggestion
          throw new IllegalArgumentException("Unable to create URL for file " + filePath, e);
```

sorry for nit-picking :-/

Reply:
Revised. Thanks a lot for your careful review.